contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+   * - To revoke all remaining revokable tokens in a schedule set amount to 0 in the `amounts` array.
+   *
+   * ##### Requirements:
+   *
+   * - Can only be used when the caller has the `TOKEN_REVOKER_ROLE`
+   * - The requirements of _beforeTokenTransfer apply to this function
+   * - toAccounts.length == removalIds.length == amounts.length
+   */
+  function batchRevokeUnreleasedTokenAmounts(
+    address[] calldata toAccounts,
+    uint256[] calldata removalIds,
+    uint256[] calldata amounts
+  ) external whenNotPaused onlyRole(TOKEN_REVOKER_ROLE) {
+    if (!(toAccounts.length == removalIds.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "removalIds"
+      });
+    }
+    if (!(toAccounts.length == amounts.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "amounts"
+      });
+    }
+    for (uint256 i = 0; i < toAccounts.length; i++) {
+      _revokeUnreleasedTokens(toAccounts[i], removalIds[i], amounts[i]);
+    }
+  }
+
+  // Private implementations ==========================================
+
+  /**
+   * Wraps minting of wrapper token and schedule setup.
+   *
+   * @dev If no schedule is set up for the specified removal id, one is created.
+   *
+   * @param removalId uint256 The removal for which funds are being deposited.
+   * @param amount uint256 Quantity of `_bridgedPolygonNori` to deposit
+   * @param userData bytes extra information provided by the user (if any)
+   * @param operatorData bytes extra information provided by the operator (if any)
+   */
+  function _depositFor(
+    uint256 removalId,
+    uint256 amount,
+    bytes memory userData,
+    bytes memory operatorData
+  ) internal returns (bool) {
+    uint256 projectId = _removal.getProjectIdForRemoval(removalId);
+    Removal.ScheduleData memory scheduleData = _removal
+      .getScheduleDataForProjectId(projectId);
+    address recipient = scheduleData.supplierAddress;
+    if (!_scheduleIdToScheduleStruct[projectId].exists) {
+      _createSchedule(projectId);
+    }
+    super._mint(recipient, projectId, amount, "");
+    Schedule storage schedule = _scheduleIdToScheduleStruct[projectId];
+    // slither-disable-next-line unused-return address may already be in set and that is ok
+    schedule.tokenHolders.add(recipient);
+    // slither-disable-next-line unused-return id may already be in the set and that is ok
+    _addressToScheduleIdSet[recipient].add(projectId);
+    return true;
+  }
+
+  /**
+   * Sets up a schedule for the specified project id (implementation).
+   *
+   * @dev schedules are created when removal tokens are listed for sale in the market contract,
+   * so this should only be invoked during `tokensReceived` in the exceptional case that
+   * tokens were sent to this contract without a schedule set up.
+   */
+  function _createSchedule(uint256 projectId) internal {
```

Comment:
this function has several storage reads that can be optimized by moving them to memory once (e.g., for `ScheduleData`)

Reply:
so I don't actually think this is true because scheduleData here is not storage, it's already memory.  I did a gas experiment between the original implementation and one that looks like this:
```
function _createSchedule(uint256 projectId) internal {
    Removal.ScheduleData memory scheduleData = _removal
      .getScheduleDataForProjectId(projectId);

    uint256 startTime = scheduleData.startTime;
    uint256 methodology = scheduleData.methodology;
    uint256 methodologyVersion = scheduleData.methodologyVersion;
    address supplierAddress = scheduleData.supplierAddress;
    if (startTime == 0) {
      revert InvalidScheduleStartTime({projectId: projectId});
    }
    address recipient = supplierAddress;
    if (recipient == address(0)) {
      revert RecipientCannotBeZeroAddress();
    }
    if (hasRole(SCHEDULE_CREATOR_ROLE, recipient)) {
      revert RecipientCannotHaveRole({
        recipient: recipient,
        role: "SCHEDULE_CREATOR_ROLE"
      });
    }
    if (!_allScheduleIds.add(projectId)) {
      revert ScheduleExists({scheduleTokenId: projectId});
    }
    Schedule storage schedule = _scheduleIdToScheduleStruct[projectId];
    uint256 restrictionDuration = getRestrictionDurationForMethodologyAndVersion(
        methodology,
        methodologyVersion
      );
    if (restrictionDuration == 0) {
      revert RestrictionDurationNotSet({projectId: projectId});
    }
    schedule.exists = _addressToScheduleIdSet[recipient].add(projectId);
    schedule.startTime = startTime;
    schedule.endTime = startTime + restrictionDuration;
    emit ScheduleCreated(projectId, schedule.startTime, schedule.endTime);
  }
```


The before (original) gas for a schedule creation is this: 
![Screen Shot 2022-06-25 at 9 59 22 AM](https://user-images.githubusercontent.com/24928141/175783672-9a0dab1f-defa-4eb7-9929-4c6ae71b7f05.png)

and after the assignment of variables for each of the struct members, it looks like this:
![Screen Shot 2022-06-25 at 10 04 08 AM](https://user-images.githubusercontent.com/24928141/175783689-6a9fb436-cee8-4d04-8cee-8e1d25a0863d.png)

Reply:
Hmm, I think might have communicated this incorrectly, what I meant is referenced elsewhere in this PR (e.g., [here](https://github.com/nori-dot-eco/contracts/pull/249/files/998a7c2f78de2bf6a1f41f2c5161c04e02449ab5..fb97bb8a727a24cdc034f908b27899c6a7b61e26#r906867575))